sdk/metric/instrument.go

```diff
@@ -24,10 +24,157 @@ import (
+	Unit unit.Unit
+	// Scope identifies the instrumentation that created the instrument.
+	Scope instrumentation.Scope
+
+	// Ensure forward compatibility if non-comparable fields need to be added.
+	nonComparable // nolint: unused
+}
+
+// mask returns a copy of base with all non-zero-value fields of m replacing
+// the fields of the returned copy.
+func (base Instrument) mask(m Instrument) Instrument {
+	cp := base
+	if m.Name != "" {
+		cp.Name = m.Name
+	}
+	if m.Description != "" {
+		cp.Description = m.Description
+	}
+	if m.Kind != zeroInstrumentKind {
+		cp.Kind = m.Kind
+	}
+	if m.Unit != zeroUnit {
+		cp.Unit = m.Unit
+	}
+	if m.Scope.Name != "" {
+		cp.Scope.Name = m.Scope.Name
+	}
+	if m.Scope.Version != "" {
+		cp.Scope.Version = m.Scope.Version
+	}
+	if m.Scope.SchemaURL != "" {
+		cp.Scope.SchemaURL = m.Scope.SchemaURL
+	}
+	return cp
+}
+
+// empty returns if all fields of i are their zero-value.
+func (i Instrument) empty() bool {
+	return i.Name == "" &&
+		i.Description == "" &&
+		i.Kind == zeroInstrumentKind &&
+		i.Unit == zeroUnit &&
+		i.Scope == zeroScope
+}
+
+// matches returns whether all the non-zero-value fields of i match the
+// corresponding fields of other. If i is empty it will match all other, and
+// true will always be returned.
+func (i Instrument) matches(other Instrument) bool {
+	return i.matchesName(other) &&
+		i.matchesDescription(other) &&
+		i.matchesKind(other) &&
+		i.matchesUnit(other) &&
+		i.matchesScope(other)
+}
+
+// matchesName returns true if the Name of i is "" or it equals the Name of
+// other, otherwise false.
+func (i Instrument) matchesName(other Instrument) bool {
+	return i.Name == "" || i.Name == other.Name
+}
+
+// matchesDescription returns true if the Description of i is "" or it equals
+// the Description of other, otherwise false.
+func (i Instrument) matchesDescription(other Instrument) bool {
+	return i.Description == "" || i.Description == other.Description
+}
+
+// matchesKind returns true if the Kind of i is its zero-value or it equals the
+// Kind of other, otherwise false.
+func (i Instrument) matchesKind(other Instrument) bool {
+	return i.Kind == zeroInstrumentKind || i.Kind == other.Kind
+}
+
+// matchesUnit returns true if the Unit of i is its zero-value or it equals the
+// Unit of other, otherwise false.
+func (i Instrument) matchesUnit(other Instrument) bool {
+	return i.Unit == zeroUnit || i.Unit == other.Unit
+}
+
+// matchesScope returns true if the Scope of i is its zero-value or it equals
+// the Scope of other, otherwise false.
+func (i Instrument) matchesScope(other Instrument) bool {
+	return (i.Scope.Name == "" || i.Scope.Name == other.Scope.Name) &&
+		(i.Scope.Version == "" || i.Scope.Version == other.Scope.Version) &&
+		(i.Scope.SchemaURL == "" || i.Scope.SchemaURL == other.Scope.SchemaURL)
+}
+
+// Stream describes the stream of data an instrument produces.
+type Stream struct {
+	Instrument
```

Comment:
Is it valid to change the InstrumentKind or Scope?

Reply:
It is unspecified. For what its worth, so is changing the unit.

I can see changing the `InstrumentKind` here as problematic, but allowing a user to fix an incorrect or poorly defined scope seems ideal. For example the instrumentation forgot to update their version or the user imports the same library twice with different names.

Reply:
I'll remove both `InstrumentKind` and `Scope` for now. I'll plan to leave `unit` though given that is something already asked for: https://github.com/open-telemetry/opentelemetry-go/discussions/3412

Reply:
Scope and unit do seem reasonable.  I wonder though if the things that can be matched should be a different struct from the things that can be updated...

Reply:
Updated.